src/storage/src/table/cell_based_table.rs

```diff
@@ -185,22 +187,103 @@ impl<S: StateStore> CellBasedTable<S> {
-        rows: Vec<(Row, Option<Row>)>,
+        buffer: BTreeMap<Row, RowOp>,
         epoch: u64,
     ) -> StorageResult<()> {
         let mut batch = self.keyspace.state_store().start_write_batch();
         let mut local = batch.prefixify(&self.keyspace);
-        for (pk, cell_values) in rows {
-            let arrange_key_buf =
-                serialize_pk(&pk, self.pk_serializer.as_ref().unwrap()).map_err(err)?;
-            let bytes = self
-                .cell_based_row_serializer
-                .serialize(&arrange_key_buf, cell_values, &self.column_ids)
-                .map_err(err)?;
-            for (key, value) in bytes {
-                match value {
-                    Some(val) => local.put(key, StorageValue::new_default_put(val)),
-                    None => local.delete(key),
+        let ordered_row_serializer = self.pk_serializer.as_ref().unwrap();
+        for (pk, row_op) in buffer {
+            let arrange_key_buf = serialize_pk(&pk, ordered_row_serializer).map_err(err)?;
+            match row_op {
+                RowOp::Insert(row) => {
+                    let bytes = self
+                        .cell_based_row_serializer
+                        .serialize(&arrange_key_buf, row, &self.column_ids)
+                        .map_err(err)?;
+                    for (key, value) in bytes {
+                        local.put(key, StorageValue::new_default_put(value))
+                    }
+                }
+                RowOp::Delete(old_value) => {
+                    // TODO(wcy-fdu): only serialize key on deletion
+                    let bytes = self
+                        .cell_based_row_serializer
+                        .serialize(&arrange_key_buf, old_value, &self.column_ids)
+                        .map_err(err)?;
+                    for (key, _) in bytes {
+                        local.delete(key);
+                    }
+                }
+                RowOp::Update((old_row, new_row)) => {
+                    let delete_bytes = self
+                        .cell_based_row_serializer
+                        .serialize(&arrange_key_buf, old_row, &self.column_ids)
+                        .map_err(err)?;
+                    let insert_bytes = self
+                        .cell_based_row_serializer
+                        .serialize(&arrange_key_buf, new_row, &self.column_ids)
+                        .map_err(err)?;
+
+                    let mut delete_idx = 0;
+                    let delete_len = delete_bytes.len();
+                    let mut insert_idx = 0;
+                    let insert_len = insert_bytes.len();
+                    loop {
+                        if delete_idx < delete_len && insert_idx < insert_len {
+                            let delete_pk = &delete_bytes[delete_idx].0;
+                            let insert_pk = &insert_bytes[insert_idx].0;
+                            let (_, delete_cell_id) = delete_pk.split_at(delete_pk.len() - 4);
```

Comment:
Define a utility function to extract column id. It's hard to read this part.

Also you may follow the suggestion of using `match` to write this part of logic. Then we also don't need an extra loop after this loop.

Reply:
I think we can also use iterator API to do this.

```rust
let insert_bytes_iter = insert_bytes.into_iter().fused().peek();
let delete_bytes_iter = delete_bytes.into_iter().fused().peek();

loop {
  match (insert_bytes_iter.peek(), delete_bytes_iter.peek()) {
    // 6 cases, call `next()` to get ownership
  }
}
```